# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: accessible/ipc/win/handler/AccessibleHandlerControl.cpp
# Commit: f262a231f046
# Full Hash: f262a231f0460c4295999b158a6b00868b49b843
# Author: James Teh <jteh@mozilla.com>
# Date: 2020-06-01 16:21:08
# Description:
#   Bug 1642141: When clearing the AccessibleHandlerControl Accessible cache, swap it into a temporary map first to avoid re-entry into the main map while clearing. r=MarcoZ
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D77623
# ==============================================================================

diff -r 2692b8d159d5 -r f262a231f046 accessible/ipc/win/handler/AccessibleHandlerControl.cpp
--- a/accessible/ipc/win/handler/AccessibleHandlerControl.cpp	Mon Jun 01 05:13:33 2020 +0000
+++ b/accessible/ipc/win/handler/AccessibleHandlerControl.cpp	Mon Jun 01 05:45:44 2020 +0000
@@ -118,7 +118,18 @@
 HRESULT
 AccessibleHandlerControl::Invalidate() {
   ++mCacheGen;
-  mAccessibleCache.clear();
+  // We can't just call mAccessibleCache.clear() because doing so would release
+  // remote objects, making remote COM calls. Since this is an STA, an incoming
+  // COM call might be handled which might marshal an AccessibleHandler,
+  // which in turn might add itself to mAccessibleCache. Since we'd be in the
+  // middle of mutating mAccessibleCache, that might cause a crash. Instead,
+  // swap mAccessibleCache into a temporary map first, which will empty
+  // mAccessibleCache without releasing remote objects. Once mAccessibleCache
+  // is empty, it's safe to let the temporary map be destroyed when it goes
+  // out of scope. Remote calls will be made, but nothing will re-enter
+  // the temporary map while it's being destroyed.
+  AccessibleCache oldCache;
+  mAccessibleCache.swap(oldCache);
   return S_OK;
 }
 